# GMI26H Cryptography Compulsory JNB Python Warm-up Lab

### You must upload your solutions in .ipynb format to Learn by Thursday September 15th at 11:59 PM with your solutions added to this file.

Write your name and email below. Don't edit the header or the questions or anything else. Add your answers in text or code to cells below the questions.

### Student name: Atichoke Nantarat 

### DU email address: H20atina@du.se

### AES Lab 1 will take place Friday September 16th at 9:00 AM sharp, so everyone is expected to have submitted their solutions to this lab by the previous night and have their JNBs running. Contact Charlie/Hasan regarding the labs if you have questions. If you cannot make the September 15th deadline, you can submit later but note that only two resubmissions are allowed. Resubmission deadlines are TBA.

## Task 1

a) All valid International Standard Book Number 13 (ISBN-13) numbers satisfy

$$m \equiv 0 \mod 10,$$

where

$$m = x_1+3x_2+x_3+3x_4+x_5+3x_6+x_7+3x_8+x_9+3x_{10}+x_{11}+3x_{12}+x_{13}$$

and $x_1$ is the first (leftmost) digit of positive integer $n$, $x_2$ is the second digit, and so on. For example, for $n=9781252434434$, $x_1=9$ and $x_2=7$, and so on.

Use Python (and/or Sage) to define a function that checks whether or not a positive integer argument $n$ is a valid ISBN-13 number.

The function should take only one argument, that is, the number $n$ (not $m$). The value of $m$ can be calculated with $n$ inside the function implementation. The function should return string "Valid" if $n$ is valid and string "Invalid" if $n$ is invalid. You can assume that $n$ is a 13-digit number.

In [11]:
#Write your code here for a)
def isbn_validator(n):
    str_n = str(n) #making the numbers into string for further use
    count = 0 #This will be used for the formular to check if it is an valid isbn number
    sum = 0 #The sum that will be % by 10 to check isbn number

    #The for loop will take the first left number addition with the ( 2nd number multiplied by 3) then repeat until the 13th number.
    for i in str_n:
        count += 1
        if count % 2 == 1:
            sum = sum +int(i)
        elif count % 2  == 0:
            k = 3*int(i)
            sum = sum + k
    #Check if ISBN number is valid
    res = sum%10
    if res == 0:
        print(f"ISBN number {n}: Valid")
    elif res >= 1:
        print(f"ISBN number {n}: Invalid")
    else:
        print("Something wrong occured.")
        print("res was : {res}")

b) Use the function that you just wrote to verify that $n=9781252434434$ is invalid.

In [12]:
#Write your code here for b)
isbn_validator(9781252434434)

ISBN number 9781252434434: Invalid


c) Use the function that you just wrote to verify that the ISBN-13 number of the course book "Stallings, W. (2022). Cryptography and Network Security: Principles and Practice, Global Edition. Pearson" is valid.

In [13]:
#Write your code here for c)
isbn_validator(9781292437484)

ISBN number 9781292437484: Valid


## Task 2

In modular arithmetic, positive integer $g$ is called a primitive root of prime $p$ if every number $a$, with $1\leq a < n$, is congruent to a power of $g$ modulo $p$.

For example, $2$ is a primitive root of $5$ because for all numbers $a$ coprime to $5$, that is, for numbers $a=1$, $a=2$, $a=3$, and $a=4$, there exists an $i$ such that a power of $2$ (a number $2^i$) is congruent to $a$ as

$$2^1=2\mod 5,$$
$$2^2=4\mod 5,$$
$$2^3=3\mod 5,$$
$$2^4=1\mod 5.$$

For $a=1$, $i=4$. For $a=2$, $i=1$. For $a=3$, $i=3$. For $a=4$, $i=2$.

a) Use Python (and/or Sage) to write a program that prints the first $12$ powers of $$11^i \mod 13$$

(where $i$ goes from $1$ to $12$) and explain why $11$ is a primitive root of $13$ by examining the printout.

In [15]:
#your code here
def mod_func():
    # 2^3 = 3, 8 (mod 5) = 3 || a = 3, i = 3
    i = 0
    a = 0
    for x in range (0,12):
        i += 1
        res = 11**i
        print(f"{res} || i:{i} ||%11:{res%13}")

    # 11^1 = 11 mod 13

mod_func()
    

11 || i:1 ||%11:11
121 || i:2 ||%11:4
1331 || i:3 ||%11:5
14641 || i:4 ||%11:3
161051 || i:5 ||%11:7
1771561 || i:6 ||%11:12
19487171 || i:7 ||%11:2
214358881 || i:8 ||%11:9
2357947691 || i:9 ||%11:8
25937424601 || i:10 ||%11:10
285311670611 || i:11 ||%11:6
3138428376721 || i:12 ||%11:1


Your explanation here: 
We can see that 11 is a primitive root of 13 with for example looking at the 2nd line which is 121 ($11^2$) and $11^2 = 4 \mod 13$
another example is 2357947691 ($11^9 = 8 \mod 13 $)

## Task 3

Let $n=pq$, where $p=43$ and $q=79$. Write a snippet of Python (and/or Sage) code that solves

$$279 = 47^i \mod n$$

for $i$. In other words, your program needs to find $i$ so that the above expression is satisfied. You can use a for-loop.

In [16]:
#your code here
# n = pq, p = 43, q = 79
i = 0
res = 0
p = 43
q = 79
n = p*q
while(True):
    if res == 279:
        print(f"i : {i}")
        #print(sum)
        break
    else:
        i += 1
        sum = 47**i
        res = sum%n



i : 38
